File: lab5/Dictionary.c
/*
* Dictionary.c
* Ly Phung
* lynphung
* 12M
* 5/11/19
* Implementation file for Dictionary.h
*/

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"

//Node
typedef struct NodeObj
{
	char* key;
	char* value;
	struct NodeObj* next;
} NodeObj;

typedef NodeObj* Node;

Node newNode(char* k, char* v)
{
	Node N = malloc(sizeof(NodeObj));
	assert(N!=NULL);
	N->key = k;
	N->value = v;
	N->next = NULL;
	return(N);
}

void freeNode(Node* pN)
{
	if(pN!=NULL && *pN!=NULL)
	{
		free(*pN);
		*pN = NULL;
	}
}

//dictionary
typedef struct DictionaryObj
{
	Node head;
	int numItems;
} DictionaryObj;

Dictionary newDictionary(void)
{
	Dictionary D = malloc(sizeof(DictionaryObj));
	assert(D!=NULL);
	D->head =NULL;
	D->numItems = 0;
	return D;
}

void freeDictionary(Dictionary* pD)
{
	if (pD!=NULL&& *pD!=NULL)
	{
		if (!isEmpty(*pD)) makeEmpty(*pD);
		free(*pD);
		*pD = NULL;
	}
}

int isEmpty(Dictionary D)
{
	if (D == NULL)
	{
		fprintf(stderr, "Dictionary Error: calling isEmpty() on a NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}

	return(D->numItems==0);
}

int size(Dictionary D)
{
	return D->numItems;
}

char* lookup(Dictionary D, char* k)
{
	if (D == NULL)
	{
		fprintf(stderr, "Dictionary Error: calling lookup() on a NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}

	Node N = D->head;
	for(;N!=NULL;N=N->next)
	{
		if (strcmp(N->key, k) ==0)
		{
			return N->value;
		}
	}
	return NULL;
}

void insert(Dictionary D, char* k, char* v)
{
	if (lookup(D, k) != NULL)
	{
		fprintf(stderr, "Dictionary Error: calling insert() on a duplicate key\n");
		exit(EXIT_FAILURE);
	}

	//if first term
	if(isEmpty(D))
	{
		D->head = newNode(k, v);
		D->numItems++;
		return;
	}

	Node N = D->head;
	for(int x = 1; x < D->numItems; x++)
	{
		N= N->next;
	}

	N->next = newNode(k, v);
	D->numItems++;
}

void delete(Dictionary D, char* k)
{
	if (lookup(D, k) == NULL)
	{
		fprintf(stderr, "Dictionary Error: calling delete() on non-existent key\n");
		exit(EXIT_FAILURE);
	}

	if(D->numItems==0)
	{
		fprintf(stderr, "Dictionary Error: calling delete() on an empty dictionary\n");
		exit(EXIT_FAILURE);
	}

	Node N = D->head;
	Node before = N;
	for(int x = 0; x < D->numItems; x++)
	{
		if(strcmp(N->key, k) == 0)
		{
			//if first term
			if(x==0)
			{
				N = D->head;
				D->head = D->head->next;
			}
			break;
		}

		before = N;
		N = N->next;
	}

	before->next = N->next;
	freeNode(&N);
	D->numItems--;
}

void makeEmpty(Dictionary D)
{
	if(D == NULL)
	{
		fprintf(stderr, "Dictionary Error: calling makeEmpty() on a NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}

	if(D->numItems==0)
	{
		fprintf(stderr, "Dictionary Error: calling makeEmpty() on an empty dictionary\n");
		exit(EXIT_FAILURE);
	}

	Node N = D->head;
	Node after = NULL;
	for(;N!=NULL;N=after)
	{
		after = N->next;
		delete(D, N->key);
	}
}

void printDictionary(FILE* out, Dictionary D)
{
	if (D ==NULL)
	{
		fprintf(stderr, "Dictionary Error: calling printDictionary() on a NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}

	Node N = D->head;
	for(;N!=NULL;N=N->next)
	{
		fprintf(out, "%s %s\n", N->key, N->value);
	}
}

File: pa3/Dictionary.java
/*------------------------------------------------------------------------------
Ly Phung 
lynphung
12B
4/27/19
Dictionary.java that creates a Dictionary ADT with the necessary operators and states.
Dictionary.java
------------------------------------------------------------------------------*/

class Dictionary implements DictionaryInterface
{

	private class Node {
      String myKey, myValue;
      Node next;

      Node(String key, String value){
      	myKey = key;
      	myValue = value;
        next = null;
      }
  	}

  	private Node head;
  	private int numItems;

  	public Dictionary()
  	{
  		head = null;
  		numItems = 0;
  	}

	public boolean isEmpty()
	{
		return (numItems ==0);
	}

	public int size()
	{
		return numItems;
	}

	public String lookup(String key)
	{
		Node N = head;
		for(;N!=null; N=N.next)
		{
			if (N.myKey.equals(key))
			{
				return N.myValue;
			}
		}
		return null;
	}

	public void insert(String key, String value) throws DuplicateKeyException
	{
		if (lookup(key)!= null)
			throw new DuplicateKeyException("cannot insert duplicate keys");
		
		//if first value
		if(head==null)
		{
			head = new Node(key, value);
			numItems++;
			return;
		}

		Node N = head;
		for(int i=1; i<numItems; i++)
		{
			N = N.next;
		}
		N.next = new Node(key, value);
		numItems++;
	}

	public void delete(String key) throws KeyNotFoundException
	{
		if (lookup(key)==null)
			throw new KeyNotFoundException("cannot delete non-existent key");

		Node N = head;
		Node before = N;
		for(int i=0; i<numItems; i++)
		{
			if (N.myKey.equals(key))
			{
				if (i==0) //if first term
				{
					head = head.next;
					N.next = null;
				}
				break;
			}
			
			before = N;
			N = N.next;
		}
		
		before.next = N.next;
		
		numItems--;
	}

	public void makeEmpty()
	{
		head = null;
		numItems=0;
	}

	public String toString()
	{
		StringBuffer line = new StringBuffer();
		Node N = head;

		for(; N!=null; N=N.next)
		{
			line.append("key=").append(N.myKey).append(" value=").append(N.myValue).append("\n");
		}

		return new String(line);
	}
}